sdk/provisioning/Azure.Provisioning/src/BicepValue.cs

```diff
@@ -2,110 +2,131 @@
 // Licensed under the MIT License.
 
 using System;
+using System.ComponentModel;
 using Azure.Provisioning.Expressions;
 using Azure.Provisioning.Primitives;
 
 namespace Azure.Provisioning;
 
-// TODO: Currently everything's wrapped in BicepValue, but we're going to pull
-// models and enums out (kind of like we do with ResourceReference already) so
-// we get a better intellisense experience.  That'll also remove a lot of the
-// clutter in this class hierarchy.
-
 /// <summary>
 /// Represents the value of a property that could be a literal .NET value, a
 /// Bicep expression, or it could be unset.
 /// </summary>
-public abstract class BicepValue
+public abstract class BicepValue : IBicepValue
```

Comment:
Just curious - why the change to an interface?

Reply:
There used to be two "hierarchies":
```mermaid
graph LR
  A[Provisonable] --> Infrastructure
  A --> B[ProvisionableConstruct]
  B --> C[NamedProvisionableConstruct]
  C --> D[Variable]
  C --> ProvisionableResource
  
  
  X[BicepValue] --> BicepValueOfT
  X --> BicepList
  X --> BicepDict 
```

This PR bridges them:

```mermaid
graph LR
  A[Provisonable] --> Infrastructure
  A --> B[ProvisionableConstruct]
  B --> C[NamedProvisionableConstruct]
  C --> D[Variable]
  C --> ProvisionableResource
  
  Z[IBicepValue] --> B
  Z --> X
    
  X[BicepValue] --> BicepValueOfT
  X --> BicepList
  X --> BicepDict 
```

The net result is now all our model types are implicitly `T | Expression` without being wrapped in `BicepValue<T>`.  This means we can replace code like `storage.PrimaryEndpoints.Value!.BlobUri` with just `storage.PrimaryEndpoints.BlobUri` because the `StorageAccountEndpoints` now knows whether it has a literal value or expression and just manages it for us.  When you start looking at ACA examples, the API was annoyingly shouty with `properties.Foo.Value!.Bar.Value!.Baz = 3` and this will make a big difference to the overall feel.